### PR TITLE
Fix sharing nested dynamic items with CatalogIndex

### LINKED ISSUE #5646: Story v2.0 - handle re-arranging catalogue not to break sharelinks (public only)
Currently re-arranging / re-organising catalogue breaks sharelinks and stories if they are shared. 
i.e. if a catalogue item is renamed or moved within a different group, it will break the share links and stories. 
reproduction of steps: 
- create a share link with a nr of data sets 
- change / reorder datasets in catalogue
- sharelink created initially will break

Does not include changing the format of datasets

### PR DESCRIPTION
**Pull request by nf-s**
### More catalog index

Fixes https://github.com/TerriaJS/terriajs/issues/5646 https://github.com/TerriaJS/product/issues/89

* Fix sharelink bug, and make `isJson*` type checks more rigorous
* Remove `uniqueId` from `CatalogMemberMixin.nameInCatalog` and add it as fallback to `CatalogMemberMixin.name`
* Add `shareKeys` and `nameInCatalog` to `CatalogIndexReference`.
* Remove `description` field from `CatalogIndex`
  * The `CatalogIndex` can now be used to resolve models in sharelinks
* Add support for zipped `CatalogIndex` json files.
* Fix `SplitReferences` which use `shareKeys`
* Make `isJson*` type assertion functions more rigorous
  * Add `deep` parameter, so you can use old "shallow" type check for performance reasons if needed
* Add `isOpen` to `TerriaReferenceTraits`

### How to test

Below is `startData` which adds the following item:
- South Australia (`TerriaReference`) -> Local government data (`CatalogGroup`) -> Burnside (`ckan-group`) -> Burnside Bike Racks (`geojson`)

I have removed all dependency (`knownContainerIds`...) information, and am referring to a sharekey for the dataset

```json
{
  "version": "8.0.0",
  "initSources": [
    {
      "stratum": "user",
      "models": {
        "Root Group/Local Government/South Australia/Burnside/0fc6b073-bdb7-4f23-9dcc-147bc3a9ae80/596efb36-1cc5-4a86-b44d-2a353f5e07e6": {
          "type": "geojson"
        }
      },
      "workbench": [
        "Root Group/Local Government/South Australia/Burnside/0fc6b073-bdb7-4f23-9dcc-147bc3a9ae80/596efb36-1cc5-4a86-b44d-2a353f5e07e6"
      ]
    }
  ]
}
```

This means that this item can move anywhere within the catalog, as long as it has the `Root Group/Local Government/South Australia/Burnside/0fc6b073-bdb7-4f23-9dcc-147bc3a9ae80/596efb36-1cc5-4a86-b44d-2a353f5e07e6` in `shareKeys` (and the `CatalogIndex` is up-to-date)

### After link

http://ci.terria.io/more-catalog-index/#configUrl=https://terria-catalogs-public.storage.googleapis.com/nationalmap/map-config/dev.json&start={%22version%22:%20%228.0.0%22,%22initSources%22:%20[{%22stratum%22:%20%22user%22,%22models%22:%20{%22Root%20Group/Local%20Government/South%20Australia/Burnside/0fc6b073-bdb7-4f23-9dcc-147bc3a9ae80/596efb36-1cc5-4a86-b44d-2a353f5e07e6%22:%20{%22type%22:%20%22geojson%22}},%22workbench%22:%20[%22Root%20Group/Local%20Government/South%20Australia/Burnside/0fc6b073-bdb7-4f23-9dcc-147bc3a9ae80/596efb36-1cc5-4a86-b44d-2a353f5e07e6%22]}]}

### Before link

http://ci.terria.io/main/#start=%7B%22version%22:%20%228.0.0%22,%22initSources%22:%20%5B%7B%22stratum%22:%20%22user%22,%22models%22:%20%7B%22Root%20Group/Local%20Government/South%20Australia/Burnside/0fc6b073-bdb7-4f23-9dcc-147bc3a9ae80/596efb36-1cc5-4a86-b44d-2a353f5e07e6%22:%20%7B%22type%22:%20%22geojson%22%7D%7D,%22workbench%22:%20%5B%22Root%20Group/Local%20Government/South%20Australia/Burnside/0fc6b073-bdb7-4f23-9dcc-147bc3a9ae80/596efb36-1cc5-4a86-b44d-2a353f5e07e6%22%5D%7D%5D%7D

### Checklist

-   [x] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
-   [ ] I've updated relevant documentation in `doc/`.
-   [x] I've updated CHANGES.md with what I changed.

### CODE REVIEW

File: buildprocess/patchNetworkRequests.ts [hunk 1]
```diff
@@ -15,83 +15,86 @@ export default function patchNetworkRequests(
...
 
-      return fetch(input as any, init as any) as any;
-    };
-    global.fetch = newFetch as any;
-  } else {
-    global.fetch = fetch as any;
-  }
+    return fetch(input as any, init as any) as any;
+  };
+  global.fetch = newFetch as any;
```
Comment (nf-s) [inline buildprocess/patchNetworkRequests.ts:97]: Note, only changes in this file are to include console logging requests

File: lib/Models/Definition/upsertModelFromJson.ts [hunk 2]
```diff
@@ -71,22 +76,13 @@ export default function upsertModelFromJson(
...
-        errors.push(
-          new TerriaError({
-            message: `Failed to get model \`"${potentialId}"\` found using share key \`"${json.id}"\``
-          })
-        );
-      }
-    }
-  }
   if (model === undefined) {
     try {
```
Comment (nf-s) [inline lib/Models/Definition/upsertModelFromJson.ts:80]: Moved to `Terria.loadModelStratum`

File: lib/Models/Definition/upsertModelFromJson.ts [hunk 3]
```diff
@@ -71,22 +76,13 @@ export default function upsertModelFromJson(
...
-      model = factory.create(json.type, uniqueId, terria);
+      model = factory.create(
+        isJsonString(json.type) ? json.type : undefined,
```
Comment (tephenavies) [inline lib/Models/Definition/upsertModelFromJson.ts]: I think it's clearer to put this ternary outside factory.create

**Review by tephenavies — CHANGES_REQUESTED** [review]

File: lib/Models/Definition/upsertModelFromJson.ts (hunk 3 above)
Comment (nf-s) [reply lib/Models/Definition/upsertModelFromJson.ts]: done!

File: lib/Models/Definition/upsertModelFromJson.ts [hunk 4]
```diff
@@ -78,11 +78,8 @@ export default function upsertModelFromJson(
 
   if (model === undefined) {
     try {
-      model = factory.create(
-        isJsonString(json.type) ? json.type : undefined,
-        uniqueId,
-        terria
-      );
+      const type = isJsonString(json.type) ? json.type : undefined;
+      model = factory.create(type, uniqueId, terria);
```
Comment (tephenavies) [inline lib/Models/Definition/upsertModelFromJson.ts]: I more meant that you know sending `type === undefined` to `factory.create` gives undefined, so why not:
```js
model = isJsonString(json.type) ? factory.create(json.type, uniqueId, terria) : undefined;
```

Comment (nf-s) [reply lib/Models/Definition/upsertModelFromJson.ts]: Oh I see. Yeah that is much better

Comment (nf-s) [reply lib/Models/Definition/upsertModelFromJson.ts]: Done!
Sorry I'm a bit slow today

**Review by tephenavies — APPROVED** [review]

**Pull request merged**